source/tutorials/learning-journey/packaging-existing-software.md

```diff
@@ -0,0 +1,531 @@
+
+For the purposes of this tutorial, "package" means something like "result of a derivation"; this is the artifact you or others will use, as a consequence of having "packaged existing software with Nix".
+:::
+
+## A Simple Project
+To start, consider this skeleton derivation:
+
+```nix
+{ stdenv }:
+
+stdenv.mkDerivation {	};
+```
+
+This is a function which takes an attribute set containing `stdenv`, and produces a derivation (which currently does nothing).
+
+As you progress through this tutorial, you will update this several times, adding more details while following the general pattern.
+
+### Hello, World!
+GNU Hello is an implementation of the "hello world" program, with source code accessible [from the GNU Project's FTP server](https://ftp.gnu.org/gnu/hello/).
+
+To begin, you will download the [latest version](https://ftp.gnu.org/gnu/hello/hello-2.12.1.tar.gz) of `hello` using `fetchzip`, which takes the URI path to the download file and a SHA256 hash of its contents.
+
+:::{note}
+`fetchzip` can fetch [more archives](https://nixos.org/manual/nixpkgs/stable/#fetchurl) than just zip files!
+:::
+
+The hash cannot be known until after the tarball has been downloaded and unpacked, but Nix will complain if the hash supplied to `fetchzip` was incorrect, so it is common practice to supply a fake one with `lib.fakeSha256` and change the derivation definition after Nix reports the correct hash:
+
+```nix
+# hello.nix
+{ lib
+, stdenv
+, fetchzip
+}:
+
+stdenv.mkDerivation {
+  src = fetchzip {
+    url = "https://ftp.gnu.org/gnu/hello/hello-2.12.1.tar.gz";
+    sha256 = lib.fakeSha256;
+  };
+}
+```
+
+Save this file to `hello.nix` and try to build it with `nix-build`, observing your first build failure:
+
+```console
+$ nix-build hello.nix
+error: cannot evaluate a function that has an argument without a value ('lib')
+       Nix attempted to evaluate a function as a top level expression; in
+       this case it must have its arguments supplied either by default
+       values, or passed explicitly with '--arg' or '--argstr'. See
+       https://nixos.org/manual/nix/stable/language/constructs.html#functions.
+
+       at /home/nix-user/hello.nix:2:3:
+
+            1| # hello.nix
+            2| { lib
+             |   ^
+            3| , stdenv
+```
+
+Problem: the expression in `hello.nix` is a *function*, which only produces its intended output if it is passed the correct *arguments*.
+
+### A New Command
+`lib` is available from `nixpkgs`, which must be imported with another Nix expression in order to pass it as an argument to this derivation.
+
+The recommended way to do this is to create a `default.nix` in the same directory as `hello.nix`, with the following contents:
+
+```nix
+# default.nix
+let
+  pkgs = import <nixpkgs> { };
+in
+{
+  hello = pkgs.callPackage ./hello.nix { };
+}
+```
+
+This allows you to use `nix-build -A hello` to realize the derivation in `hello.nix`, similar to the current convention used in `nixpkgs`.
```

Comment:
I don't disagree, but it's not *too* much more to type and we eventually want people to help maintain the broader project.

Reply:
I agree with @proofconstruction, it was also me who originally said that we don't want to use `nix-build -E 'with import <nixpkgs> {}; callPackage ./hello.nix { }`, because that's already an antipattern in Nixpkgs.

And the alternative of using just a single buildable file like this:

```nix
with import <nixpkgs> {};
stdenv.mkDerivation {
  ...
}
```

has the problem that it doesn't support `.override` (`callPackage` adds that), and it would be different from the Nixpkgs convention. It would also mean that if you have multiple packages, Nixpkgs would get imported multiple times and in potentially different ways and different versions.

Reply:
I'm not going to block on this suggestion, I just think it's important to remember that most users don't care about doing things the `nixpkgs` way because it doesn't provide any value _to them_.

Reply:
I think there's a lot of value in doing it like this, independently of Nixpkgs:
- It cleanly separates the definition of the package (`hello.nix`) from the importing of its dependencies (Getting some Nixpkgs path in `default.nix`). This is just good practice in general from a separation-of-concerns point-of-view, but it also allows reusing the files in various ways, e.g. you could call the same `hello.nix` with different versions of dependencies, or you could copy the `hello.nix` to `hello-alt.nix` and change it to get a different build with the same dependencies.
- `callPackage` adds the `.override` functionality, which allows easy overriding of dependencies and custom package configuration, fairly clearly separated from the implementation.

But still, we cannot deny Nixpkgs existing, it adds more benefits on top:
- All packages in Nixpkgs are packaged using this very same convention, allowing users to either copy package expressions directly from Nixpkgs, modify and build them, or the other way around, easily upstream packages to Nixpkgs.
- All packages in Nixpkgs have a `.override`, and people are used to being able to change dependencies and configuration using it.